coderd/database/dbauthz/dbauthz.go

```diff
@@ -1278,38 +1278,62 @@ func (q *querier) customRoleEscalationCheck(ctx context.Context, actor rbac.Subj
 	return nil
 }
 
+// requireSystemRestrictedActor checks to make sure the caller is a system restricted actor.
+func (q *querier) requireSystemRestrictedActor(ctx context.Context, action policy.Action, object rbac.Object) error {
+	act, ok := ActorFromContext(ctx)
+	if !ok {
+		return ErrNoActor
+	}
+
+	if act.Type == rbac.SubjectTypeSystemRestricted {
+		return nil
+	}
+
+	// We use a standard RBAC forbidden error to avoid leaking details
+	// to clients while still being verbose in tests/logs.
+	internal := xerrors.New("invalid actor, system restricted actor required")
+	err := rbac.ForbiddenWithInternal(internal, act, action, object, nil)
+	return logNotAuthorizedError(ctx, q.log, err)
+}
+
 // customRoleCheck will validate a custom role for inserting or updating.
 // If the role is not valid, an error will be returned.
 // - Check custom roles are valid for their resource types + actions
 // - Check the actor can create the custom role
 // - Check the custom role does not grant perms the actor does not have
-// - Prevent negative perms
-// - Prevent roles with site and org permissions.
+// - Prevent negative perms for non-system roles
+// - Prevent roles that have both (site|user) and org permissions.
```

Comment:
```suggestion
// - Prevent roles that have both organization scoped and non-organization scoped permissions.
```

lets not use syntax in comments like this. the point of comments is to describe things plainly where syntax is not enough. 😄